CMakeLists.txt

```diff
@@ -1,504 +1,2030 @@
+  message (FATAL_ERROR "Builds (in-source) are not allowed in "
+    "${CMAKE_CURRENT_SOURCE_DIR}. Please remove CMakeCache.txt and the CMakeFiles "
+    "directory from ${CMAKE_CURRENT_SOURCE_DIR} and try building in a separate directory.")
+endif ()
+
+if ("${CMAKE_GENERATOR}" MATCHES "Visual Studio" AND
+    NOT ("${CMAKE_GENERATOR}" MATCHES .*Win64.*))
+  message (FATAL_ERROR
+    "Visual Studio 32-bit build is not supported. Use -G\"${CMAKE_GENERATOR} Win64\"")
+endif ()
 
-############################################################
+if (NOT CMAKE_SIZEOF_VOID_P EQUAL 8)
+  message (WARNING "Rippled requires a 64 bit target architecture.\n"
+    "The most likely cause of this warning is trying to build rippled with a 32-bit OS.")
+endif ()
+
+#[===================================================================[
+   read version from source
+#]===================================================================]
+
+file (STRINGS src/ripple/protocol/impl/BuildInfo.cpp BUILD_INFO)
+foreach (line_ ${BUILD_INFO})
+  if (line_ MATCHES "versionString[ ]*=[ ]*\"(.+)\"")
+    set (rippled_version ${CMAKE_MATCH_1})
+  endif ()
+endforeach ()
+if (rippled_version)
+  message (STATUS "rippled version: ${rippled_version}")
+else ()
+  message (FATAL_ERROR "unable to determine rippled version")
+endif ()
+
+#[===================================================================[
+   declare user options/settings
+#]===================================================================]
+
+option (assert "Enables asserts, even in release builds" OFF)
+option (unity "Creates a build based on unity sources. This is the default" ON)
+if (NOT WIN32)
+  option (coverage "Generates coverage info." OFF)
+else ()
+  set (coverage OFF CACHE BOOL "unix only" FORCE)
+endif ()
+if (is_gcc)
+  option (profile "Add profiling flags" OFF)
+else ()
+  set (profile OFF CACHE BOOL "GCC only" FORCE)
+endif ()
+if (is_linux)
+  option (static "link protobuf, openssl, libc++, and boost statically" OFF)
+  option (perf "Enables flags that assist with perf recording" OFF)
+  option (use_gold "enables detection of gold (binutils) linker" ON)
+else ()
+  set (static OFF CACHE BOOL "static link, linux only" FORCE)
+  set (perf OFF CACHE BOOL "perf flags, linux only" FORCE)
+  set (use_gold OFF CACHE BOOL "gold linker, linux only" FORCE)
+endif ()
+option (jemalloc "Enables jemalloc for heap profiling" OFF)
+option (werr "treat warnings as errors" OFF)
+
+# this one is a string and therefore can't be an option
+set (san "" CACHE STRING "On gcc & clang, add sanitizer instrumentation")
+set_property (CACHE san PROPERTY STRINGS ";undefined;memory;address;thread")
+if (san)
+  string (TOLOWER ${san} san)
+  if (NOT (is_gcc OR is_clang))
+    message (FATAL_ERROR "san option only works for clang/gcc")
+  endif ()
+  if (san STREQUAL "memory" AND NOT (is_clang AND
+    CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7))
+    message (FATAL_ERROR "memory sanitizer only available for clang 7+")
```

Comment:
Is this message meant to apply to AppleClang 7 or higher, or LLVM 7 (unreleased yet)? I would assume the former since I think even LLVM 4 or so has MSAN.

Reply:
good point - this is a weak check anyhow. I'm going to convert this to `check_cxx_compiler_flag` which is more directly what we want

Reply:
Agreed, leveraging `check_cxx_compiler_flag` is a good call.